zanata-client-commands/src/main/java/org/zanata/client/commands/push/RawPushCommand.java

```diff
@@ -75,10 +94,164 @@ public RawPushCommand(PushOptions opts, RestClientFactory clientFactory) {
+            for (DocumentType docType: acceptedTypes) {
+                for (String extension: docType.getSourceExtensions()) {
+                    if (inputExtensions.contains(extension)) {
+                        String msg = "Please use file type '"
+                            + docType
+                            + "' for extension '"
+                            + extension
+                            + "', e.g. --file-types "
+                            + docType;
+                        throw new RuntimeException(msg);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Return list of validated type with list of extensions
+     *
+     * Validate user input file types against server accepted file types
+     *
+     * - Each file type is only input once
+     * - Same file extension cannot be in multiple file type
+     * - Source file extension that is not supported from file type will be ignored.
+     *
+     * eg. input pattern --file-types properties_utf8,properties[txt],plain_text[md;markdown],html[txt],idml
+     *
+     * @param acceptedTypes
+     * @param inputFileTypes
+     */
+    private Map<DocumentType, Set<String>> validateFileTypes(
+            List<DocumentType> acceptedTypes, List<String> inputFileTypes) {
+
+        Map<DocumentType, Set<String>> filteredFileTypes = new HashMap<>();
+
+        for (String typeWithExtension : inputFileTypes) {
+            String fileType = extractFileType(typeWithExtension);
+            List<String> extensions = extractExtensions(typeWithExtension);
+
+            validateInputFileType(fileType, extensions, acceptedTypes);
+
+            DocumentType inputFileType = DocumentType.getByName(fileType);
+
+            //skip file type if its not supported
+            if (inputFileType == null) {
+                String msg =
+                        "Requested file type '"
+                                + fileType
+                                + "' is not supported by the target server and will be ignored.";
+                consoleInteractor.printfln(DisplayMode.Warning, msg);
+                continue;
+            }
+
+            //throw error if file type is input more than once
+            if (filteredFileTypes.containsKey(inputFileType)) {
+                String msg =
+                    "File type '" + fileType + "' cannot be used more than once in option --file-types.";
+                log.error(msg);
+                throw new RuntimeException(msg);
+            }
+
+            Set<String> filteredExtensions = new HashSet<>(extensions);
+
+            if (!extensions.isEmpty()) {
+                //throw exception if user input extension if not part of DocumentType supported source extension
+                for (String extension: extensions) {
+                    if (!inputFileType.getSourceExtensions()
+                            .contains(extension)) {
```

Comment:
If we don't know the extension, then we can just use the same extension for source and translation. That is all users will want most of the time.

For the rare case that someone does need a custom extension that is different between source and translation, it is ok for that case not to work. I think it will be possible later when we update the File project type to handle different directory structures.